plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -26,13 +26,18 @@ interface TypeEmbedding {
 
     fun accessInvariants(v: Exp): List<Exp> = emptyList()
 
+    // This is a list of invariants that are already known to be true, thus they are just assumed by Viper instead of explicitly proven.
+    // An example of this is when other systems (e.g. the type checker) have already proven these.
+    fun provenInvariants(v: Exp): List<Exp> = emptyList()
+
     fun invariants(v: Exp): List<Exp> = emptyList()
 
     /**
      * Invariants that should correlate the old and new value of a value of this type
-     * over a function call.  This is exclusively necessary for CallsInPlace.
-     *
-     * TODO: add support for these in loops, too.
```

Comment:
I think this TODO is still relevant, or does your earlier loop invariants stuff already do that?  I thought there would be a problem, actually, since loops don't support `old`...

Reply:
I don't think the TODO is really relevant anymore. As far as I understood it `old` is supported in loops but just doesn't refer to the head state of the previous loop iteration but rather the start of the method. However, that is kind of exactly what we want since `dynamicInvariants` should express the things we want to have at the end of the method call. Putting them into the loop invariants allows us to prove them.

Reply:
Yeah a tricky thing is also that `old` only works for heap-dependent things (like field accesses). If you tried it with variables that doesn't work. The semantics are a bit weird.